src/Microsoft.Health.Dicom.Core/Features/CustomTag/IndexableDicomTagService.cs

```diff
@@ -0,0 +1,68 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dicom;
+using EnsureThat;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Dicom.Core.Configs;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Query;
+
+namespace Microsoft.Health.Dicom.Core.Features.CustomTag
+{
+    public class IndexableDicomTagService : IIndexableDicomTagService
+    {
+        private readonly ICustomTagStore _customTagStore;
+        private readonly bool _enableCustomQueryTags;
+        public static readonly IReadOnlyList<IndexableDicomTag> CoreIndexableDicomTags = GetCoreIndexableDicomTags();
+        private List<IndexableDicomTag> _allIndexableDicomTags;
+
+        public IndexableDicomTagService(ICustomTagStore customTagStore, IOptions<FeatureConfiguration> featureConfiguration)
+        {
+            EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
+            EnsureArg.IsNotNull(featureConfiguration?.Value, nameof(featureConfiguration));
+            _customTagStore = customTagStore;
+            _enableCustomQueryTags = featureConfiguration.Value.EnableCustomQueryTags;
+        }
+
+        public async Task<IReadOnlyCollection<IndexableDicomTag>> GetIndexableDicomTagsAsync(CancellationToken cancellationToken = default)
+        {
+            if (_enableCustomQueryTags)
+            {
+                if (_allIndexableDicomTags == null)
+                {
+                    _allIndexableDicomTags = new List<IndexableDicomTag>(CoreIndexableDicomTags);
+
+                    IReadOnlyList<CustomTagStoreEntry> customTagEntries = await _customTagStore.GetCustomTagsAsync(cancellationToken: cancellationToken);
+                    foreach (CustomTagStoreEntry customTagEntry in customTagEntries)
+                    {
+                        DicomTag tag = DicomTag.Parse(customTagEntry.Path);
+                        DicomVR vr = DicomVR.Parse(customTagEntry.VR);
+                        _allIndexableDicomTags.Add(new IndexableDicomTag(tag, vr, customTagEntry.Level, isCustomTag: true));
+                    }
+                }
+
+                return _allIndexableDicomTags;
+            }
+            else
+            {
+                return CoreIndexableDicomTags;
+            }
+        }
+
+        private static IReadOnlyList<IndexableDicomTag> GetCoreIndexableDicomTags()
```

Comment:
Is'nt this already available with AllInstanceTags?

Reply:
Yes, but need to convert.
I feel this might be used globally just like AllInstanceTags, so make it as "public static readonly".

Reply:
Why not move all these conversions to IndexTag and you can do AllInstanceTags.Select(t => Convert(t)).ToList();

Reply:
The extenstion method will simplify, I'll do that.
While IndexableDicomTag also requires the tagLevel which need to setup manually, so I'm thinking of still keeping this method, how do you think?